### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by sz16900**
# **Milestone 3: Game logic**

## **Suggested branch name game_logic.**

- Add real logic for your game - get rid of all inline comments and mocked logic used in previous milestone.
- As main.rb file should be responsible for user interaction, this is the only(!) one place where puts/print and get.chomp can be used.
- Think how you can encapsulate diffrent parts of game logic in classes. Do not create one class for everything.
- Read about DRY rule and use it.
- Think about the diffrence between attr_readers/attr_writers and atrr_acessors. Use atrr_acessors only if you really need it.
- Remember about single responsibility rule. Always.
- Remember to cover edge cases: think what player could try to do to cheat the game and make sure that your program does not allow to perform invalid action.

## **Our Improvements**

- Added scoreboard.
- Added infinite game loop until prompted to end game.

### CODE REVIEW

File: lib/display.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    (0..2).each do |row|
+      print '|'
+      (0..2).each do |col|
+        print board[row][col]
+        print '|'
+      end
+      puts ''
+    end
+    puts ''
+  end
```
Comment (euqueme) [inline lib/display.rb]: - [ ] All puts, prints and gets chomps should be in the main.rb file, all methods should return only objects (strings, integers, booleans, boards, games) 

For example: here you could store the board in a string and return the string and just call puts print_board(board) somewhere within your main.rb file

File: lib/display.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    print 'Choose a number between 1 and 9: '
+    x = gets.chomp
+    raise 'An Error' unless ('1'..'9').include? x
+
+    @position = x.to_i
+  rescue StandardError
+    puts ''
+    puts 'Wrong Input! Please try again.'
+    retry
+  end
```
Comment (euqueme) [inline lib/display.rb]: - [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). this one is not only printing and requesting for input is also validating, move all the printing and request for input to the main.rb

File: lib/display.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  def player_wins(player)
+    puts ''
+    puts "Player #{player} is the WINNER!!!"
+  end
+
+  def draw
+    puts ''
+    puts "It's a DRAW!!!"
+  end
```
Comment (euqueme) [inline lib/display.rb]: - [ ] All puts, prints and gets chomps should be in the main.rb file, all methods should return only objects (strings, integers, booleans, boards, games) so please modify your position_taken, player_wins and draw methods

File: lib/display.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    elsif yes == 'n'
+      true
+    else
+      raise 'An error'
+    end
+  rescue StandardError
+    puts ''
+    puts 'Wrong Input! Please try again.'
+    retry
+  end
```
Comment (euqueme) [inline lib/display.rb]: - [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). score_board and play_again? are not only printing and/or requesting for input they are also validating, move all the printing and request for input to the main.rb

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  private
+
+  def end_game_handler
+    @display.print_board(@board.the_board)
+    @display.score_board(@scores)
+    @game_over = @display.play_again?
+    @board.reset_board
+  end
```
Comment (euqueme) [inline lib/game.rb]: - [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). start and end_game_handler have many responsibilities

Tip: I think these methods could be avoided you only need a loop in your main.rb with most of the logic of these methods, let's work on it

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @sz16900  and @fdezcaminero  :balloon: :lollipop: 

### First of all :clap: claps to you for having a descriptive Readme file at this point, and for the descriptive pull request :bowing_woman:  it will definitely be less to work on for the next milestone :medal_sports: :medal_sports: secondly, your project works, is playable and you've done a good work keeping your code dry.
 
### there is still some work :construction: :construction_worker_man: :construction_worker_man:  to do and I have made comments on your code with the changes that you need to make so please address them bellow :arrow_double_down: 

### Also address this

- [ ]  whenever the player decides to play again, Player O always starts the game regardless if he/she lost the previous match.

### That's it, team! :raised_hands: 

### Please :pray: do the changes :man_technologist: :man_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @sz16900 and @fdezcaminero :handshake: :handshake: 

### Thank you for doing the changes I requested in the first review, :medal_sports: :medal_sports:  your project looks more organized, I can see you created a loop in your main file which is the ideal here :ok_hand: there is still some work to do:

## Implementation details

#### single responsibility rule for methods

- [ ] The methods you have within your main file (win_method, draw_method and end_game_handler) are still not following the single responsibility rule, you are printing and also retrieving values, move all the printing to the loop so your methods have one single responsibility.

## unexpected behavior:

- [ ] when a player wins, the game is asking the next player to make a move when it should display the winning message right away

image

- [ ] when a player wins with the last move the game is counting it as a draw instead of a win for the player

image

### That's it, team! :raised_hands: 

### Please :pray: do the changes :man_technologist: :man_technologist: and submit back for a code review when you are done :memo: 

### Happy coding! :rocket: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Review by pelzolga123 — APPROVED** [review]
Approved!

**Comment by pelzolga123** [comment]
## Great  work, Congratulation :+1: Project Approved.
You did all changes required previously. Go ahead and merge to development.
_Goodluck in your next milestone_ *Happy coding*

**Pull request merged**